feat: add support for phpstan>=2.1.12

maxwellmezadre opened the pull request:
Since [PHPStan 2.1.12](https://github.com/phpstan/phpstan/releases/tag/2.1.12) introduces an additional constructor parameter in NodeScopeResolver, this PR adjusts the constructor invocation for PHPStan ≥ 2.1.12.

![CleanShot 2025-04-16 at 17 14 55](https://github.com/user-attachments/assets/04260eb6-d0f5-40bd-a4e9-243e063bb87b)

maxwellmezadre committed:
feat: add support for phpstan>=2.1.12

nimah79 commented:
But it needs another argument after the other arguments (`narrowMethodScopeFromConstructor`).

I think it'll be much better if the `NodeScopeResolver` is instantiated using named arguments. No need to check PHPStan version:

```php
$nodeScopeResolver = new NodeScopeResolver( // @phpstan-ignore-line
    reflectionProvider: $reflectionProvider,
    initializerExprTypeResolver: $container->getByType(InitializerExprTypeResolver::class),
    reflector: $container->getService('betterReflectionReflector'), // @phpstan-ignore-line
    classReflectionExtensionRegistryProvider: $container->getByType(ClassReflectionExtensionRegistryProvider::class),
    parameterOutTypeExtensionProvider: $container->getByType(ParameterOutTypeExtensionProvider::class),
    parser: $container->getService('defaultAnalysisParser'), // @phpstan-ignore-line
    fileTypeMapper: $container->getByType(FileTypeMapper::class),
    stubPhpDocProvider: $container->getByType(StubPhpDocProvider::class),
    phpVersion: $container->getByType(PhpVersion::class),
    signatureMapProvider: $container->getByType(SignatureMapProvider::class),
    deprecationProvider: $container->getByType(DeprecationProvider::class),
    attributeReflectionFactory: $container->getByType(AttributeReflectionFactory::class),
    phpDocInheritanceResolver: $container->getByType(PhpDocInheritanceResolver::class),
    fileHelper: $container->getByType(FileHelper::class),
    typeSpecifier: $typeSpecifier, // @phpstan-ignore-line
    dynamicThrowTypeExtensionProvider: $container->getByType(DynamicThrowTypeExtensionProvider::class),
    readWritePropertiesExtensionProvider: $container->getByType(ReadWritePropertiesExtensionProvider::class),
    parameterClosureTypeExtensionProvider: $container->getByType(ParameterClosureTypeExtensionProvider::class),
    scopeFactory: $scopeFactory,
    polluteScopeWithLoopInitialAssignments: false,
    polluteScopeWithAlwaysIterableForeach: true,
    polluteScopeWithBlock: true, // @phpstan-ignore-line
    earlyTerminatingMethodCalls: [],
    earlyTerminatingFunctionCalls: [],
    universalObjectCratesClasses: [], // @phpstan-ignore-line
    implicitThrows: true,
    treatPhpDocTypesAsCertain: true,
    narrowMethodScopeFromConstructor: true,
);
```

@maxwellmezadre

maxwellmezadre committed:
refactor: add named arguments

maxwellmezadre commented:
I’ve switched the NodeScopeResolver instantiation to named arguments and added the extra narrowMethodScopeFromConstructor parameter as requested.
However, the PHPStan version check has to stay—other contributors and CI jobs may still be running different PHPStan releases, so removing the guard would break their builds. Keeping the check lets the code work reliably across the full range of supported versions.

@nimah79

ondrejmirtes commented:
NodeScopeResolver constructor isn't covered by BC promise (https://phpstan.org/developing-extensions/backward-compatibility-promise) because it often changes.

The proper solution is to fetch NodeScopeResolver via `Container::getByType()`.

pelmered commented:
> NodeScopeResolver constructor isn't covered by BC promise (https://phpstan.org/developing-extensions/backward-compatibility-promise) because it often changes.
> 
> The proper solution is to fetch NodeScopeResolver via `Container::getByType()`.

Do you mean that all this code including all the version checks could be replaced with: 
`$nodeScopeResolver = Container::getByType(NodeScopeResolver::class)`

maxwellmezadre committed:
refactor: improve readability and maintainability

maxwellmezadre commented:
I’ve applied the improvements focusing on readability and maintainability.

Here’s how the code looks now — I’m retrieving the required services (NodeScopeResolver and FileAnalyser) using $container->getByType(), as suggested. I also moved the rule and collector registry creation into clearly named variables for better clarity.

Thanks a lot for the great tips — they helped clean this up nicely!

@ondrejmirtes
@owenvoke

maxwellmezadre committed:
refactor: improve readability and maintainability

maxwellmezadre committed:
refactor: improve readability and maintainability

nimah79 commented on src/PHPStanAnalyser.php:
```diff
@@ -4,168 +4,35 @@
 {
     /**
-     * Creates an analyser with the rules and collectors needed for type coverage.
+     * Creates an analyzer with the rules and collectors needed for type coverage.
```
Why? the class name is `PHPStanAnalyser` (it uses `s` in `analyser`)

owenvoke replied:
Yeah, this change can be reverted as it's unnecessary. The other bits look good to me. 👍🏻

maxwellmezadre committed:
Merge remote-tracking branch 'origin/3.x' into 3.x

maxwellmezadre committed:
refactor: revert fix description

pelmered commented:
Nice cleanup!

maxwellmezadre committed:
Merge branch '3.x' into 3.x
